provider/resources.go

```diff
@@ -989,12 +989,31 @@ func Provider() tfbridge.ProviderInfo {
 					Source: "compute_backend_service_signed_url_key.html.markdown",
 				},
 			},
-			"google_compute_disk":                          {Tok: gcpResource(gcpCompute, "Disk")},
-			"google_compute_firewall":                      {Tok: gcpResource(gcpCompute, "Firewall")},
-			"google_compute_firewall_policy":               {Tok: gcpResource(gcpCompute, "FirewallPolicy")},
-			"google_compute_firewall_policy_rule":          {Tok: gcpResource(gcpCompute, "FirewallPolicyRule")},
-			"google_compute_firewall_policy_association":   {Tok: gcpResource(gcpCompute, "FirewallPolicyAssociation")},
-			"google_compute_forwarding_rule":               {Tok: gcpResource(gcpCompute, "ForwardingRule")},
+			"google_compute_disk":                        {Tok: gcpResource(gcpCompute, "Disk")},
+			"google_compute_firewall":                    {Tok: gcpResource(gcpCompute, "Firewall")},
+			"google_compute_firewall_policy":             {Tok: gcpResource(gcpCompute, "FirewallPolicy")},
+			"google_compute_firewall_policy_rule":        {Tok: gcpResource(gcpCompute, "FirewallPolicyRule")},
+			"google_compute_firewall_policy_association": {Tok: gcpResource(gcpCompute, "FirewallPolicyAssociation")},
+			"google_compute_forwarding_rule": {
+				Tok: gcpResource(gcpCompute, "ForwardingRule"),
+				TransformFromState: func(_ context.Context, pMap resource.PropertyMap) (resource.PropertyMap, error) {
+					// The service_directory_registrations field changes from [] in v6 to an object in v7.
+					// It appears that on early versions of v7, upstream has state migrators; they are lost by v7.38
+					// of this provider. See https://github.com/pulumi/pulumi-gcp/issues/2471.
+					pValue := resource.NewProperty(pMap)
+					serviceDirRegistrationsPath := resource.PropertyPath{"serviceDirectoryRegistrations"}
+					val, ok := serviceDirRegistrationsPath.Get(pValue)
+					if ok && val.IsArray() {
+						if len(val.ArrayValue()) > 0 {
```

Comment:
I don't believe this branch is covered by tests, can we please add a test for a non-nil serviceDirectoryRegistrations?
You can use https://github.com/pulumi/providertest/pull/95 like in https://github.com/pulumi/pulumi-azure/blob/12949f95dacdda96ffd9d7710d73bc1b48bda565/provider/provider_yaml_test.go#L79